R/simple_ensemble.R

```diff
@@ -0,0 +1,140 @@
+#' Compute ensemble model outputs by summarizing component model outputs for
+#' each combination of model task, output type, and output type id. Supported
+#' output types include `mean`, `median`, `quantile`, `cdf`, and `pmf`.
+#'
+#' @param model_outputs an object of class `model_output_df` with component
+#'   model outputs (e.g., predictions).
+#' @param weights an optional `data.frame` with component model weights. If
+#'   provided, it should have columns `model_id`, `weight`, and optionally,
+#'   additional columns corresponding to task id variables, `output_type`, or
+#'   `output_type_id`, if weights are specific to values of those variables. The
+#'   default is `NULL`, in which case an equally-weighted ensemble is calculated.
+#' @param agg_fun a function or character string name of a function to use for
+#'   aggregating component model outputs into the ensemble outputs. See the
+#'   details for more information.
+#' @param agg_args a named list of any additional arguments that will be passed
+#'   to `agg_fun`.
+#' @param model_id `character` string with the identifier to use for the
+#'   ensemble model.
+#' @param task_id_cols, output_type_col, output_type_id_col, value_col
+#'   `character` vectors with the names of the columns in `model_outputs` for
+#'   the output's type, additional identifying information, and value of the
+#'   model output.
+#'
+#' @details The default for `agg_fun` is `mean`, in which case the ensemble's
+#'   output is the average of the component model outputs within each group
+#'   defined by a combination of values in the task id columns, output type, and
+#'   output type id. The provided `agg_fun` should have an argument `x` for the
+#'   vector of numeric values to summarize, and for weighted methods, an
+#'   argument `w` with a numeric vector of weights. For weighted methods,
+#'  `agg_fun = "mean"` and `agg_fun = "median"` are translated to use
+#'  `matrixStats::weightedMean` and `matrixStats::weightedMedian` respectively.
+#'
+#' @return a data.frame with columns `team_abbr`, `model_abbr`, one column for
+#'   each task id variable, `output_type`, `output_id`, and `value`. Note that
+#'   any additional columns in the input `model_outputs` are dropped.
+simple_ensemble <- function(model_outputs, weights = NULL,
+                            agg_fun = "mean", agg_args = list(),
+                            model_id = "hub-ensemble",
+                            task_id_cols = NULL,
+                            output_type_col = "output_type",
+                            output_type_id_col = "output_type_id",
+                            hub_connection = NULL) {
+  model_out_cols <- colnames(model_outputs)
```

Comment:
Probably best to move the `colnames()` call to after checking `model_outputs` is a data.frame